### PR TITLE
Replace per-node measure functions with a single measure function + per-node context

### PR DESCRIPTION
**Pull request by nicoburns**
# Objective

- Make it possible for measure functions to borrow external state rather than being required to own all of the state they need.
- Make the `Send+Sync` bounds on measure functions opt-in

## Changes made

A new parameter (of user-specifiable generic type) has been added to:

- The `Taffy::compute_layout` function
- The `measure` function of the `Measurable` trait (measure functions)
- The Send + Sync bound has been removed from the Measurable trait
- A new `SyncMeasureFunc` type with Send+Sync bounds has been added

The parameter passed to the `Taffy::compute_layout` function by the user is passed through when Taffy calls into measure functions.

To make this work, an associated type `Context` has been added to the `Measurable` trait, and a generic type parameter `Measure: Measurable` has been added to the `Taffy` struct. This generic parameter also allows users of Taffy to choose between the Send+Sync `SyncMeasureFunc` and the regular `MeasureFunc` which doesn't have those bounds, or to implement their own custom `Measurable` type (allowing for things like using an enum instead of a boxed closure).

## Context

- Will allow https://github.com/trimental/inlyne to remove unnecessary Mutex's from their Taffy integration
- Will allow WASM bindings to use Taffy without a sketchy unsafe Send+Sync implementation
- Will allow Bevy to create a custom `Measureable` type and avoid boxed closures

## Feedback wanted

- Does this seem like a good direction?
- Context is currently always passed by mutable reference. Can we do better?
- How should we handle constructors? Taffy having a generic parameter is mostly painless, but it does make constructing a `Taffy` more awkward.

### CODE REVIEW

**Comment by nicoburns** [comment]
@twop I would be interested on your opinions on this design given your work in #198

File: src/tree/measure_func.rs
Comment (TimJentzsch) [inline src/tree/measure_func.rs:69]: Typo:
```suggestion
    /// Call the measure function to measure the node
```

File: src/tree/taffy_tree/tree.rs
Comment (TimJentzsch) [inline src/tree/taffy_tree/tree.rs:441]: I feel like this function could cause unintended side-effects because the `self.context` is overridden.
Maybe it could make sense to save the old `self.context` in a variable and then restore that instead of overwriting with `None`?

In any case, I think the behavior with `self.context` should be documented in the comment.

**Review by TimJentzsch — COMMENTED** [review]
At first glance, I don't mind this implementation and it seems to unlock a fair bit of use cases.

Because you have to commit to one `Context` for everything I was first concerned that it might be harder to create reuseable functionality for this system. But then I'm not sure if reuse across multiple taffy client would even make sense or how that could be implemented differently... so it probably doesn't matter.

I left some minor comments about typos and documentation.

File: README.md
Comment (nicoburns) [reply README.md:23]: I can try it, but I'm guessing not on the basis that `Measure` defaults to `MeasureFunc<()>` so if that were possible then it ought to be possible to just write `Taffy` (but I tried that and it didn't work). I'm thinking maybe an alias like `type DefaultTree = Taffy<MeasureFunc<()>>`?

See also: https://github.com/DioxusLabs/taffy/issues/449#issuecomment-1572935359

**Comment by twop** [comment]
> @twop I would be interested on your opinions on this design given your work in https://github.com/DioxusLabs/taffy/pull/198

@nicoburns I love the idea! I think it solves some paint points of the sketch I made in https://github.com/DioxusLabs/taffy/pull/198, but it is spiritually the same what I had in mind. I think your implementation is more elegant :)

I'm not sure about the exact API ergonomics, for example 
```rs
 pub fn compute_layout_with_context(
        &mut self,
        node: NodeId,
        available_space: Size<AvailableSpace>,
        context: Measure::Context,
    ) -> Result<(), TaffyError> {
        self.context = Some(context);
        let result = compute_layout(self, node, available_space);
        self.context = None;
        result
    }
```
feels intuitively a bit off, specifically that pattern.
```rs
   self.context = Some(context);
   self.context = None;
```
But I don't know on top of my head how it can be structured a bit better.


Overall, I think that direction is a big win in my head for the library (as a lib user). Nice job 👍!!

**Comment by twop** [comment]
One other interesting detail/observation: `Context` needs to be `mut` for `measure` call (for example for caching, thus changing internal state), so `compute_layout_with_context` takes ownership of `Context` to potentially mutate it internally. I wonder if it is worth giving a shot passing `&mut Context` instead, it can be potentially more ergonomic for implementers. 

Maybe I'm missing something, but it seems that using the API properly might involve some `Rc<RefCell>` voodoo, at least in some cases, thus `&mut Context` with potentially some lifetime tracking might provide better dev ergonomics

**Comment by inobelar** [comment]
@nicoburns I'm sorry that I'm "climbing into someone else's monastery with my charter" (I'm a C++ programmer, and I only "read" Rust, and don't "write"), but doesn't instantiating a `Taffy` template/generic with different types generate twice as much code? For example:
```rust
let mut taffy1: Taffy<MeasureFunc<TextLayoutData>> = Taffy::new(); // Instantiation for one type
let mut taffy2: Taffy<MeasureFunc<SomeOtherData>> = Taffy::new();  // Instantiation for another type
```
Two different instances produces 2 unique `Taffy` specializations & double code generation (if in Rust it works same as in C++). 
I know, that its unexpected/unusal code example (2 separated intances of nodes tree), but even if we have only single instance for application, and we need extra data during measuring widget size - can we avoid the template parameter for the whole `Taffy` tree structure (since it may be various types in different cases)? 

For example, in C++, not too-safe, but good-old common-pattern for passing any context - using non-owning `void` pointer (type erasurement from known type), and then back - casting into known type when it's needed. For example (C++-like pseudocode):
```cpp
class SomeSpeciallyMeasuredWidget
{
    taffy::NodeId _id;
    taffy::Style _style;
    
    SomeUsefulDataForMeasurement _meas_data;
    
    SomeSpeciallyMeasuredWidget (Widget* parent)
         : Widget(parent)
    {
         taffy::Taffy* taffy_tree = this->getRootTree();
         // `this` is mutable pointer to self, here is it passed as `void*`
         _id = taffy_tree->new_leaf_with_measure_and_context(_style, this, 
             [](const Size<Option<float>>& known_dimensions, const Size<taffy::AvailableSpace>& available_space, void* context) -> taffy::Size<float>
             {
                 SomeSpeciallyMeasuredWidget* self = static_cast<SomeSpeciallyMeasuredWidget*>(context); // from void* into known type
                 return my_measurement_computation(known_dimensions, available_space, self->_meas_data);
             }
         );
    }
};
```

Maybe somehow we can achieve the same in Rust?

**Comment by twop** [comment]
@inobelar I think your intuition is correct, but I'm not sure what is the impact of the issue in practice.

E.g. How many different `Measure` implementations any given project uses? Judging from personal subjective experience, I don't see a common usage pattern where that is the case 🤷‍♀️

**Comment by nicoburns** [comment]
@inobelar You're definitely correct that this will generate two copies of everything. And this definitely seems non-ideal given that it's only a very small part of the code that would actually need to change.

The Rust equivalent of `void *` in this case would probably be `&mut dyn Any`: that can be downcast to `&mut dyn T` if you know the correct T. I'm a little reluctant to make that change though, as it would require all measure function implementations (written by the user) to manually downcast the value.

The other way to avoid duplicating everything would be to make the algorithms take `&mut dyn LayoutTree` rather than `&mut impl LayoutTree`. But I'm also reluctant to do that as I suspect that might cause performance issues.

The expected way of using Taffy would be:

```rust

struct TextLayoutData { ... }
struct SomeOtherData { ... }

enum CombinedData {
     Text(TextLayoutData),
     Other(SomeOtherData),
}

let mut taffy: Taffy<MeasureFunc<CombinedData>> = Taffy::new();
```

which also avoids this problem. And I agree with @twop that it would be uncommon to need multiple Taffy instances with different Context parameters.

File: src/tree/taffy_tree/tree.rs
Comment (twop) [reply src/tree/taffy_tree/tree.rs:441]: That is the same conclusion I had when I experimented with this idea, so that's why I was super excited when your approach allowed to preserve the same API surface area (more or less). 

But I think fundamentally there needs to be another implementation of `LayoutTree` for more a "advanced" use case.

**Comment by lucasmerlin** [comment]
I'm building a [integration](https://github.com/lucasmerlin/hello_egui/tree/main/crates/egui_taffy) to use taffy for egui for more complex layouts. I based my work on this PR since I hoped it would solve the lifetime issue I was facing, but in the end I had to use some unsafe code to make things work. 


<details><summary>My api basically looks like this: </summary>
<p>


```rust
    let mut taffy = TaffyPass::new(
        ui,
        Id::new("flexible"),
        Style {
            display: Display::Flex,
            ..Default::default()
        },
    );

    taffy.add(
        Id::new("child_2"),
        Style {
            flex_grow: 1.0,
            ..Default::default()
        },
        Layout::centered_and_justified(Direction::TopDown),
        |ui| {
            let _ = ui.button("Button 1");
        },
    );

    taffy.add(
        Id::new("child_3"),
        Style {
            flex_grow: 1.0,
            ..Default::default()
        },
        Layout::centered_and_justified(Direction::TopDown),
        |ui| {
            let _ = ui.button("Button 2");
        },
    );
    taffy.show();
```

</p>
</details> 


Since egui is immediate mode and I'm creating this as a library to be used in combination with egui's layout capabilities, my use of taffy is probably a bit different than most others. But I think what I'm trying to do should be possible with safe rust.

My problem is: The last parameter to the add function is a FnMut closure. I want to pass this as a &mut reference to the measure function and then after the layout was calculated, I want to call it again to create the actual ui. But since the compute_layout_with_context requires the context to live for the length of the Taffy struct, this is not possible.

Here is a simple example that fails due to a similar lifetime problem:

<details><summary>Details</summary>
<p>

```rust

use taffy::prelude::{Size, Style};
use taffy::tree::MeasureFunc;
use taffy::Taffy;

struct TaffyState<'f> {
    taffy: Taffy<MeasureFunc<&'f mut str>>,
}

fn main() {
    let mut state = TaffyState {
        taffy: Taffy::new(),
    };

    let node = state
        .taffy
        .new_with_children(Style { ..Style::default() }, &[])
        .unwrap();

    {
        let mut str = "hello".to_string();

        let result = state
            .taffy
            .compute_layout_with_context(node, Size::length(100.0), &mut str)
            .unwrap();
    }
}

```

</p>
</details> 

For reference, [here](https://github.com/lucasmerlin/hello_egui/blob/8cc499727436626c85555156adc0ca582a52d213/crates/egui_taffy/src/lib.rs#L289) is the unsafe code I had to write to make things work. I think what I'm doing there should be sound, but I'm not very experienced with rust safety.

Even if this won't be possible for some reason, your PR makes what I'm trying to do much easier, so I really appreciate it!

**Comment by nicoburns** [comment]
Hmm… the lifetime being required to outlive the Taffy struct doesn’t sound right! I suspect this is a result of the hack (temporarily storing the context in the Taffy struct for the duration of the layout) that is currently being used to avoid having to pass the context around.

I suspect I’m going to have to update this PR to do things properly, which will likely mean creating a TaffyView struct or similar which can act as an “iterator” over the Taffy tree without actually being the tree. We should be able to store the context in that struct without causing this lifetime issue.

I’m quite busy over the next month, so I’m not sure when I’ll next have time to work on Taffy. But this PR is top of my list, and the change discussed above is the main thing blocking this PR from being merged, so hopefully it shouldn’t be too long :)

**Comment by inobelar** [comment]
@nicoburns , again I apologize for the offtopic, but adding of this new 'generic' parameter also will affect the 'C bindings' https://github.com/DioxusLabs/taffy/pull/404 and 'Wasm bindings' https://github.com/DioxusLabs/taffy/pull/394 PRs, or any 'bindings to other languages' in general. 

Since, we cannot expose generics through 'C ABI' - we need some 'specialization' for that case. Is it will be: `Taffy<MeasureFunc<()>>`, or something different?

**Comment by nicoburns** [comment]
C and WASM bindings are actually one of the motivations for this PR. It allows them to define their own type to use as the measure function. For C, I am imagining this being something like:

```rust
struct CMeasureFunc {
    context: *mut (),
    measure: fn(
        known_dimensions: Size<Option<f32>>,
        available_space: Size<AvailableSpace>,
        context: *mut ()
     ) -> Size<f32>,
}
```

where `*mut ()` is a Rust representation of a C `void *` pointer. (there will probably need to be some tweaks to the types of the `known_dimensions` and `available_space` parameters too, in order to avoid using data-containing enums).

For WASM, there would be some type that wraps a handle to a JavaScript function. Notably such functions are `!Send` and therefore won't work well with the regular `MeaureFunction`.

File: README.md
Comment (nicoburns) [reply README.md:23]: My latest thinking on this is that we ought to rename the generic struct to `TaffyTree`, and then have a `type Taffy = TaffyTree<MeasureFunc<()>>` for backwards compatibility, but otherwise encourage users of Taffy to define their own alias, and documenting a bunch of helpful examples on the `TaffyTree` struct.

**Comment by nicoburns** [comment]
@lucasmerlin This will likely undergo further revision before it's merged. But the latest version of this PR solves your lifetime issue :)

Your example can now be written:

```rust
use taffy::prelude::{Size, Style};
use taffy::tree::MeasureFunc;
use taffy::Taffy;

struct TaffyState {
    taffy: Taffy<MeasureFunc<str>>,
}

fn main() {
    let mut state = TaffyState {
        taffy: Taffy::new(),
    };

    let node = state
        .taffy
        .new_with_children(Style { ..Style::default() }, &[])
        .unwrap();

    {
        let mut str = "hello".to_string();

        let result = state
            .taffy
            .compute_layout_with_context(node, Size::length(100.0), &mut str)
            .unwrap();
    }
}
```

**Comment by lucasmerlin** [comment]
@nicoburns I gave the changes a try and it's definitely an improvement. Unfortunately I'm still running into lifetime problems. 
I'm storing the Taffy struct in a TypeMap in egui's context, and my MeasureFunc basically has the context type Box<dyn FnMut() + 'a>. But since my Taffy Struct is stored in a TypeMap, 'a basically can only be 'static. 


<details><summary>This example should illustrate this:</summary>
<p>

```rust
use std::sync::{Arc, Mutex};
use taffy::prelude::{Size, Style};
use taffy::tree::MeasureFunc;
use taffy::Taffy;

type ContentFn<'a> = Box<dyn FnMut() + Send + 'a>;

struct TaffyState<'a> {
    taffy: Taffy<MeasureFunc<Vec<ContentFn<'a>>>>,
}

struct TaffyPass<'a> {
    // I'm storing this in a type_map in egui's context, so there is no way (that I know of) to
    // give this a lifetime other than 'static.
    state: Arc<Mutex<TaffyState<'static>>>,
    content_fns: Vec<ContentFn<'a>>,
}

fn main() {
    // State persists across frames, so it's stored in egui's context.
    let mut state = Arc::new(Mutex::new(TaffyState {
        taffy: Taffy::new(),
    }));

    // Every frame I create a new pass which holds the content fns
    let mut pass = TaffyPass {
        state: state.clone(),
        content_fns: Vec::new(),
    };

    pass.layout();
}

impl<'a> TaffyPass<'a> {
    fn layout(mut self) {
        let mut state = self.state.lock().unwrap();
        let node = state
            .taffy
            .new_with_children(Style { ..Style::default() }, &[])
            .unwrap();

        let result = state
            .taffy
            .compute_layout_with_context(node, Size::length(100.0), &mut self.content_fns)
            .unwrap();

        println!("{:?}", result);
    }
}
```

</p>
</details> 

I think in theory this could be solved if the MeasureFunc could be passed to the compute_layout_with_context, instead of passing it to the nodes individually, which would mean the Taffy struct wouldn't be generic anymore, solving this lifetime problem. 

I imagine something like this:
```rust
    let result = state.taffy.compute_layout_with_measure(node, Size::length(100.0), |node: NodeId, known_size, available_size| {
        ...
    });
```

The taffy struct could also still be generic over some sort of NodeData type, so there'd be some way to store some data in a taffy node, which could then also be passed to the new measure function.

I know this would be a huge api change but from my perspective this seems to be the most flexible api.

**Comment by nicoburns** [comment]
@ickshonpe Do you have any thoughts on the API proposed by @lucasmerlin above? Or indeed any of the originally proposed changes on this PR.

I was hoping to avoid such a drastic API change. But I can't see how to solve the lifetime issue any other way, so I think we might have to go that way!

**Comment by nicoburns** [comment]
@lucasmerlin I've implemented more or less the API you suggested above. I have included the generic context type (and currently a node *must* have a context set in order for the measure function to be called on that node). Usage is as follows:

```rust
struct MyContextType { ... };

let mut tree = Taffy<MyContextType>::new();
let node = tree.new_leaf_with_context(Style::default(), MyContextType { ... });
tree.compute_layout_with_measure(
    node,
    Size::MAX_CONTENT,
    |known_dimensions: Size<Option<f32>>, available_space: Size<AvailableSpace>, node_id: NodeId, context: Option<&mut MyContextType>| {
          ...
    }
).unwrap()
```

You can use `()` for the context if you don't need one. Or an enum if you have multiple different kinds of leaf node.

Technically with this approach there doesn't need to be a "context" at all (as users could use the `NodeId` to lookup per-node data externally). But it does seem potentially quite useful. Would be interested on people's thoughts on whether it should be kept.

**Comment by lucasmerlin** [comment]
@nicoburns Awesome! I was able to remove the unsafe workaround from egui_taffy with these changes. Works perfectly.

I agree that the context is useful, I think it makes sense to keep it.

**Comment by vkurchatkin-miro** [comment]
@nicoburns kudos, this is an amazing change!

> Will allow WASM bindings to use Taffy without a sketchy unsafe Send+Sync implementation

That is exactly a problem that I have. I was expecting that integration text measurements would be a smooth sail, but had to resort to not just `Rc<RefCell>`, but also this, which seems to be only required to enable multithreaded layout which is not possible in WASM anyways

**Comment by nicoburns** [comment]
@vkurchatkin-miro I’m glad that this helps! I feel like the current version of this PR seems like good design, so hopefully we’ll get it merged and then in a stable release soon! Out of interest, what are you using Taffy for? (no worries if you don’t want to share)

@ali

**Comment by nicoburns** [comment]
@alice-i-cecile @TimJentzsch A re-review on this would be amazing if you have the time. I feel like this is almost ready to be merged now, and probably just needs a final round of review. Sorry there have been so many iterations on this PR!

**Comment by TimJentzsch** [comment]
> How should we handle constructors? Taffy having a generic parameter is mostly painless, but it does make constructing a Taffy more awkward.

What would you think about making `Taffy::new` non-generic and always returning `Taffy::<()>`, and then adding a new function for creating the generic instances?
I think that could reduce the number of changes in the PR a lot and reduce boiler plate for the simple cases, but perhaps makes the generic cases less discoverable.

**Comment by vkurchatkin-miro** [comment]
@nicoburns so far just some experiments with building WASM-based backend for React

**Comment by nicoburns** [comment]
> What would you think about making `Taffy::new` non-generic and always returning `Taffy::<()>`, and then adding a new function for creating the generic instances? I think that could reduce the number of changes in the PR a lot and reduce boiler plate for the simple cases, but perhaps makes the generic cases less discoverable.

Hmm... I think my take is that:

```rust
let tree : Taffy<()> = Taffy::new();
```
isn't that much worse than
```rust
let tree = Taffy::new();
```

Especially considering that:

- Most apps will only ever construct one `Taffy` instance.
- Most apps will probably actually want to use context (although with this PR you can also use `NodeId`s with borrowed context - and I'm not sure which approach will be more popular).

Having said that, I wouldn't object to making this change if others think it's better.

Does anyone else have an opinion on this? @alice-i-cecile ?

**Comment by alice-i-cecile** [comment]
I'm alright with the existing API: I don't think ergonomics here are important enough to make the change.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Motivation is great, the new example is very helpful, and this seems like a reasonable way to achieve some rather important goals.

Like always, code quality is excellent. I'm happy to merge and ship this. However, we should update the release notes in this PR, especially since it's breaking.

**Comment by nicoburns** [comment]
> However, we should update the release notes in this PR, especially since it's breaking.

Ah, good point. I'll get on that!

**Comment by nicoburns** [comment]
@alice-i-cecile Release notes added :)

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**
